lib/Feature/FeatureModelParser.cpp

```diff
@@ -233,4 +234,318 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+      }
+
+      // Add a child
+      OrGroup = OrGroupMapping.find(CurrentIndentationLevel - 1);
+      if (OrGroup != OrGroupMapping.end()) {
+        std::get<2>(OrGroup->second).push_back(Name);
+      }
+
+
+      LastIndentationLevel = CurrentIndentationLevel;
+    }
+
+    // Add the remaining or groups
+    for (auto & OrGroup : OrGroupMapping) {
+      FMB.emplaceRelationship(std::get<1>(OrGroup.second), std::get<2>(OrGroup.second), std::get<0>(OrGroup.second));
+    }
+  }
+
+  return true;
+}
+
+bool FeatureModelSxfmParser::parseConstraints(xmlChar *Constraints) {
+  // TODO (se-passau/VaRA#702): This has to wait until the constraint part is implemented
+  return true;
+}
+
+std::optional<std::tuple<int, int>> FeatureModelSxfmParser::extractCardinality(const string& StringToExtractFrom) {
+  std::optional<int> MinCardinality;
+  std::optional<int> MaxCardinality;
+
+  // Search for the first occurrence of '['; then read in the min cardinality
+  // until the comma. Afterwards, read in the max cardinality until ']'
+  std::string::size_type Pos = StringToExtractFrom.find_first_of('[');
+  if (Pos == std::string::npos) {
+    std::cerr << "No cardinality given in or group!" << std::endl;
+    return std::optional<std::tuple<int, int>>();
+  }
+  auto CardinalityString = llvm::StringRef(StringToExtractFrom);
+  size_t CommaPos = CardinalityString.find(',', Pos + 1);
+  MinCardinality = parseCardinality(CardinalityString.substr(Pos + 1, CommaPos - Pos - 1).str());
+  Pos = CommaPos;
+  MaxCardinality = parseCardinality(CardinalityString.substr(Pos + 1, CardinalityString.find(']', Pos + 1) - Pos - 1).str());
+
+  if (!MinCardinality.has_value() || !MaxCardinality.has_value()) {
+    std::cerr << "No parsable cardinality!" << std::endl;
+    return std::optional<std::tuple<int, int>>();
+  }
+
+  if (MinCardinality.value() != 1 || (MaxCardinality.value() != 1 && MaxCardinality.value() != UINT_MAX)) {
+    std::cerr << "Cardinality unsupported. We support cardinalities [1,1] (alternative) or [1, *] (or group)." << std::endl;
+    return std::optional<std::tuple<int, int>>();
+  }
+
+  return std::optional<std::tuple<int, int>>(std::tuple<int, int>{MinCardinality.value(), MaxCardinality.value()});
+}
+
+std::optional<int> FeatureModelSxfmParser::parseCardinality(const string& CardinalityString) {
+  std::optional<int> Result = std::optional<int>();
+  if (CardinalityString == "*") {
+    // We use UINT_MAX as our magic integer (which is -1 as int) to indicate
+    // that the cardinality is a wildcard.
+    Result = UINT_MAX;
+  } else {
+    // Convert the string into an integer in a safe way
+    char *End;
+    long LongNumber;
+    errno = 0;
+    LongNumber = strtol(CardinalityString.c_str(), &End, 0);
+    if (errno == ERANGE || LongNumber < INT_MIN || LongNumber > INT_MAX
+        || *CardinalityString.c_str() == '\0' || *End != '\0') {
+      std::cerr << "The following cardinality is not integer: " << CardinalityString << std::endl;
```

Comment:
```suggestion
    if (!llvm::to_integer(CardinalityString, LongNumber, 10)) {
      llvm::errs() << llvm::formatv("The cardinality: '{0}' was not an integer.\n", CardinalityString);
```
llvm has us already covered to make converting strings to numbers easier :smiley: